File: backend/app/schemas/product.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.enums import ProductFieldType
from app.schemas.product_category import ProductCategorySummary

MAX_PRODUCT_IMAGES = 8


class ProductImageResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    image_url: str
    thumbnail_url: str | None
    alt_text: str | None
    sort_order: int
    mime_type: str | None
    width: int | None
    height: int | None


class ProductImageInput(BaseModel):
    image_url: str = Field(min_length=1, max_length=500)
    thumbnail_url: str | None = Field(default=None, max_length=500)
    alt_text: str | None = Field(default=None, max_length=255)
    sort_order: int = 0
    mime_type: str | None = Field(default=None, max_length=100)
    width: int | None = Field(default=None, ge=1)
    height: int | None = Field(default=None, ge=1)


class ProductImageReorderRequest(BaseModel):
    ordered_ids: list[int] = Field(min_length=1)


class ProductVariantInput(BaseModel):
    """Seller-facing input for a product variant (roadmap task 16)."""

    name: str = Field(min_length=1, max_length=255)
    price_override: Decimal | None = Field(default=None, gt=0)
    stock_quantity: int = Field(default=0, ge=0)
    sort_order: int = 0
    is_active: bool = True


class ProductVariantResponse(ProductVariantInput):
    model_config = ConfigDict(from_attributes=True)

    id: int
    product_id: int
    created_at: datetime
    updated_at: datetime


class ProductFieldOption(BaseModel):
    label: str = Field(min_length=1, max_length=255)
    value: str = Field(min_length=1, max_length=255)


class ProductFormFieldInput(BaseModel):
    field_key: str = Field(min_length=1, max_length=100)
    label: str = Field(min_length=1, max_length=255)
    field_type: ProductFieldType
    sort_order: int = 0
    is_required: bool = False
    placeholder: str | None = Field(default=None, max_length=255)
    help_text: str | None = None
    validation: dict[str, str | int | float | bool | None] | None = None
    options: list[ProductFieldOption] | None = None

    @model_validator(mode="after")
    def validate_options(self) -> "ProductFormFieldInput":
        if self.field_type in {ProductFieldType.DROPDOWN, ProductFieldType.RADIO} and not self.options:
            raise ValueError("Dropdown and radio fields require options")
        if self.field_type not in {ProductFieldType.DROPDOWN, ProductFieldType.RADIO} and self.options:
            # Keep the payload tidy; non-select inputs should not carry stale options.
            self.options = None
        return self


class ProductFormFieldResponse(ProductFormFieldInput):
    model_config = ConfigDict(from_attributes=True)

    id: int
    product_id: int
    created_at: datetime
    updated_at: datetime


class ProductFormFieldReorderRequest(BaseModel):
    ordered_ids: list[int] = Field(min_length=1)


class ProductCreate(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    description: str | None = None
    price: Decimal = Field(gt=0)
    stock_quantity: int = Field(ge=0, default=0)
    is_active: bool = True
    video_url: str | None = Field(default=None, max_length=500)
    video_mime_type: str | None = Field(default=None, max_length=100)
    image_urls: list[str] | None = Field(default=None, max_length=MAX_PRODUCT_IMAGES)
    images: list[ProductImageInput] | None = Field(default=None, max_length=MAX_PRODUCT_IMAGES)
    form_fields: list[ProductFormFieldInput] | None = None
    variants: list[ProductVariantInput] | None = None
    category_id: int | None = None
    shipping_cost: Decimal | None = Field(default=None, ge=0)


class ProductUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    price: Decimal | None = Field(default=None, gt=0)
    stock_quantity: int | None = Field(default=None, ge=0)
    is_active: bool | None = None
    video_url: str | None = Field(default=None, max_length=500)
    video_mime_type: str | None = Field(default=None, max_length=100)
    image_urls: list[str] | None = Field(default=None, max_length=MAX_PRODUCT_IMAGES)
    images: list[ProductImageInput] | None = Field(default=None, max_length=MAX_PRODUCT_IMAGES)
    form_fields: list[ProductFormFieldInput] | None = None
    variants: list[ProductVariantInput] | None = None
    category_id: int | None = None
    shipping_cost: Decimal | None = Field(default=None, ge=0)


class ProductResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    store_id: int
    title: str
    description: str | None
    price: Decimal
    stock_quantity: int
    is_active: bool
    video_url: str | None
    video_mime_type: str | None
    images: list[ProductImageResponse]
    form_fields: list[ProductFormFieldResponse]
    variants: list[ProductVariantResponse] = Field(default_factory=list)
    category_id: int | None = None
    category: ProductCategorySummary | None = None
    shipping_cost: Decimal | None = None
    created_at: datetime
    updated_at: datetime

File: frontend/types/seller/product.ts
export type ProductFieldType =
  | "TEXT"
  | "TEXTAREA"
  | "NUMBER"
  | "DROPDOWN"
  | "RADIO"
  | "CHECKBOX"
  | "FILE_UPLOAD";

export type ProductFieldOption = {
  label: string;
  value: string;
};

export type ProductImage = {
  id: number;
  image_url: string;
  thumbnail_url: string | null;
  alt_text: string | null;
  sort_order: number;
  mime_type: string | null;
  width: number | null;
  height: number | null;
};

export type ProductImageInput = {
  image_url: string;
  thumbnail_url?: string | null;
  alt_text?: string | null;
  sort_order?: number;
  mime_type?: string | null;
  width?: number | null;
  height?: number | null;
};

export type ProductFormField = {
  id: number;
  product_id: number;
  field_key: string;
  label: string;
  field_type: ProductFieldType;
  sort_order: number;
  is_required: boolean;
  placeholder: string | null;
  help_text: string | null;
  validation: Record<string, string | number | boolean | null> | null;
  options: ProductFieldOption[] | null;
  created_at: string;
  updated_at: string;
};

export type ProductFormFieldInput = {
  field_key: string;
  label: string;
  field_type: ProductFieldType;
  sort_order?: number;
  is_required?: boolean;
  placeholder?: string | null;
  help_text?: string | null;
  validation?: Record<string, string | number | boolean | null> | null;
  options?: ProductFieldOption[] | null;
};

export type ProductVariant = {
  id: number;
  product_id: number;
  name: string;
  price_override: string | null;
  stock_quantity: number;
  sort_order: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
};

export type ProductVariantInput = {
  name: string;
  price_override?: number | string | null;
  stock_quantity?: number;
  sort_order?: number;
  is_active?: boolean;
};

export type Product = {
  id: number;
  store_id: number;
  title: string;
  description: string | null;
  price: string;
  stock_quantity: number;
  is_active: boolean;
  video_url?: string | null;
  video_mime_type?: string | null;
  images: ProductImage[];
  form_fields: ProductFormField[];
  variants: ProductVariant[];
  created_at: string;
  updated_at: string;
};

export type ProductCreate = {
  title: string;
  description?: string | null;
  price: number | string;
  stock_quantity?: number;
  is_active?: boolean;
  video_url?: string | null;
  video_mime_type?: string | null;
  image_urls?: string[] | null;
  images?: ProductImageInput[] | null;
  form_fields?: ProductFormFieldInput[] | null;
  variants?: ProductVariantInput[] | null;
};

export type ProductUpdate = {
  title?: string;
  description?: string | null;
  price?: number | string;
  stock_quantity?: number;
  is_active?: boolean;
  video_url?: string | null;
  video_mime_type?: string | null;
  image_urls?: string[] | null;
  images?: ProductImageInput[] | null;
  form_fields?: ProductFormFieldInput[] | null;
  variants?: ProductVariantInput[] | null;
};

export type ProductImageReorderRequest = {
  ordered_ids: number[];
};

export type ProductFormFieldReorderRequest = {
  ordered_ids: number[];
};
